File: api/schemas/iq_agent.py
"""
from __future__ import annotations

IQ Agent Schemas for GraphRAG Compliance Intelligence

Pydantic schemas for IQ agent API endpoints including:
- Compliance queries and analysis requests
- GraphRAG response structures
- Memory management operations
- Risk assessment and reporting
"""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import Field

from api.schemas.base import BaseSchema, StandardResponse

class ComplianceQueryRequest(BaseSchema):
    """Request schema for compliance queries to IQ agent"""

    query: str = Field(
        ...,
        description="Natural language compliance query",
        min_length=1,
        max_length=2000,
        example="What are the GDPR compliance gaps for our data processing activities?",
    )
    context: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Additional context for the query (regulations, business functions, etc.)",
        example={
            "business_functions": ["data_processing", "customer_onboarding"],
            "regulations": ["GDPR", "DPA2018"],
            "risk_tolerance": "high",
        },
    )
    include_graph_analysis: bool = Field(
        default=True,
        description="Whether to include detailed graph analysis in response",
    )
    include_recommendations: bool = Field(
        default=True, description="Whether to include actionable recommendations",
    )

class GraphContext(BaseSchema):
    """Graph analysis context from IQ's GraphRAG processing"""

    nodes_traversed: int = Field(..., description="Number of graph nodes analyzed")
    patterns_detected: List[Dict[str, Any]] = Field(
        ..., description="Compliance patterns detected during analysis",
    )
    memories_accessed: List[str] = Field(..., description="IDs of relevant memories accessed")
    learnings_applied: int = Field(..., description="Number of learning insights applied")

class ComplianceSummary(BaseSchema):
    """High-level compliance status summary"""

    risk_posture: str = Field(..., description="Overall risk posture", example="HIGH")
    compliance_score: float = Field(
        ..., description="Overall compliance coverage score (0.0-1.0)", ge=0.0, le=1.0,
    )
    top_gaps: List[str] = Field(..., description="Top compliance gaps identified", max_items=5)
    immediate_actions: List[str] = Field(..., description="Immediate actions required", max_items=5)

class ActionPlan(BaseSchema):
    """Individual action in compliance plan"""

    action_id: str = Field(..., description="Unique action identifier")
    action_type: str = Field(..., description="Type of action")
    target: str = Field(..., description="Action target/objective")
    priority: str = Field(..., description="Action priority level")
    regulation: str = Field(..., description="Related regulation code")
    risk_level: str = Field(..., description="Associated risk level")
    cost_estimate: float = Field(..., description="Estimated implementation cost")
    timeline: str = Field(..., description="Estimated timeline")
    graph_reference: str = Field(..., description="Reference to graph node/relationship")

class ComplianceArtifacts(BaseSchema):
    """Detailed compliance analysis artifacts"""

    compliance_posture: Dict[str, Any] = Field(
        ..., description="Detailed compliance posture analysis",
    )
    action_plan: List[ActionPlan] = Field(..., description="Prioritized action plan")
    risk_assessment: Dict[str, Any] = Field(..., description="Risk assessment details")

class ComplianceEvidence(BaseSchema):
    """Evidence of compliance controls and execution"""

    controls_executed: int = Field(..., description="Number of controls executed")
    evidence_stored: int = Field(..., description="Number of evidence items stored")
    metrics_updated: int = Field(..., description="Number of metrics updated")

class NextAction(BaseSchema):
    """Next recommended action"""

    action: str = Field(..., description="Action description")
    priority: str = Field(..., description="Priority level")
    owner: str = Field(..., description="Responsible party")
    graph_reference: str = Field(..., description="Graph reference")

class IQAgentResponse(BaseSchema):
    """Complete IQ agent response with GraphRAG analysis"""

    status: str = Field(..., description="Response status")
    timestamp: str = Field(..., description="Response timestamp")

    summary: ComplianceSummary = Field(..., description="High-level compliance summary")
    artifacts: ComplianceArtifacts = Field(..., description="Detailed compliance artifacts")
    graph_context: GraphContext = Field(..., description="Graph analysis context")
    evidence: ComplianceEvidence = Field(..., description="Evidence and execution details")
    next_actions: List[NextAction] = Field(..., description="Recommended next actions", max_items=5)
    llm_response: str = Field(..., description="Natural language response from IQ")

class MemoryStoreRequest(BaseSchema):
    """Request to store memory in IQ's knowledge base"""

    memory_type: str = Field(
        ..., description="Type of memory to store", example="compliance_insight",
    )
    content: Dict[str, Any] = Field(..., description="Memory content to store")
    importance_score: float = Field(
        default=0.5, description="Importance score for memory retention", ge=0.0, le=1.0,
    )
    tags: Optional[List[str]] = Field(default=None, description="Tags for memory categorization")

class MemoryRetrievalRequest(BaseSchema):
    """Request to retrieve memories from IQ's knowledge base"""

    query: str = Field(..., description="Query for memory retrieval")
    max_memories: int = Field(
        default=10, description="Maximum number of memories to retrieve", ge=1, le=50,
    )
    relevance_threshold: float = Field(
        default=0.5, description="Minimum relevance threshold", ge=0.0, le=1.0,
    )

class MemoryNode(BaseSchema):
    """Individual memory node structure"""

    id: str = Field(..., description="Memory identifier")
    memory_type: str = Field(..., description="Type of memory")
    content: Dict[str, Any] = Field(..., description="Memory content")
    timestamp: datetime = Field(..., description="Memory creation timestamp")
    importance_score: float = Field(..., description="Importance score")
    access_count: int = Field(..., description="Number of times accessed")
    tags: List[str] = Field(..., description="Memory tags")
    confidence_score: float = Field(..., description="Confidence score")

class MemoryRetrievalResponse(BaseSchema):
    """Response from memory retrieval operation"""

    query_id: str = Field(..., description="Query identifier")
    retrieved_memories: List[MemoryNode] = Field(..., description="Retrieved memory nodes")
    relevance_scores: List[float] = Field(
        ..., description="Relevance scores for retrieved memories",
    )
    total_memories_searched: int = Field(..., description="Total number of memories searched")
    retrieval_strategy: str = Field(..., description="Strategy used for retrieval")
    confidence_score: float = Field(..., description="Overall confidence in results")

class GraphInitializationRequest(BaseSchema):
    """Request to initialize compliance graph"""

    clear_existing: bool = Field(default=False, description="Whether to clear existing graph data")
    load_sample_data: bool = Field(
        default=True, description="Whether to load sample compliance data",
    )

class GraphInitializationResponse(BaseSchema):
    """Response from graph initialization"""

    status: str = Field(..., description="Initialization status")
    timestamp: str = Field(..., description="Initialization timestamp")
    nodes_created: Dict[str, int] = Field(..., description="Count of nodes created by type")
    relationships_created: int = Field(..., description="Number of relationships created")
    message: str = Field(..., description="Status message")

class HealthCheckResponse(BaseSchema):
    """IQ agent health check response"""

    status: str = Field(..., description="Agent status")
    neo4j_connected: bool = Field(..., description="Neo4j connection status")
    graph_statistics: Dict[str, Any] = Field(..., description="Graph database statistics")
    memory_statistics: Dict[str, Any] = Field(..., description="Memory system statistics")
    last_query_time: Optional[datetime] = Field(
        default=None, description="Time of last query processed",
    )

# Response type aliases for FastAPI
ComplianceQueryResponse = StandardResponse[IQAgentResponse]
MemoryStoreResponse = StandardResponse[Dict[str, str]]
MemoryRetrievalResponseWrapper = StandardResponse[MemoryRetrievalResponse]
GraphInitializationResponseWrapper = StandardResponse[GraphInitializationResponse]
IQHealthCheckResponse = StandardResponse[HealthCheckResponse]

File: frontend/types/iq-agent.ts
/**
 * IQ Agent TypeScript Types
 * 
 * Complete type definitions matching the sophisticated GraphRAG backend schemas
 * and documented Postman API collection endpoints.
 */

// Base Response Types
export interface BaseResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

// ===== COMPLIANCE QUERY TYPES =====

export interface IQComplianceQueryRequest {
  query: string;
  context?: {
    business_functions?: string[];
    regulations?: string[];
    risk_tolerance?: 'low' | 'medium' | 'high';
    [key: string]: any;
  };
  include_graph_analysis?: boolean;
  include_recommendations?: boolean;
}

export interface ComplianceSummary {
  risk_posture: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  compliance_score: number; // 0.0 to 1.0
  top_gaps: string[];
  immediate_actions: string[];
}

export interface ActionPlan {
  action_id: string;
  action_type: string;
  target: string;
  priority: 'low' | 'medium' | 'high' | 'critical';
  regulation: string;
  risk_level: string;
  cost_estimate: number;
  timeline: string;
  graph_reference: string;
}

export interface ComplianceArtifacts {
  compliance_posture: {
    overall_coverage: number;
    total_gaps: number;
    critical_gaps: number;
    high_risk_gaps: number;
    [key: string]: any;
  };
  action_plan: ActionPlan[];
  risk_assessment: {
    convergence_patterns: number;
    recent_regulatory_changes: number;
    overall_risk_level: string;
    planning_horizon: string;
    [key: string]: any;
  };
}

export interface GraphContext {
  nodes_traversed: number;
  patterns_detected: Array<{
    pattern_type: string;
    confidence: number;
    description: string;
    implications: string[];
    [key: string]: any;
  }>;
  memories_accessed: string[];
  learnings_applied: number;
  [key: string]: any;
}

export interface ComplianceEvidence {
  controls_executed: number;
  evidence_stored: number;
  metrics_updated: number;
}

export interface NextAction {
  action: string;
  priority: string;
  owner: string;
  graph_reference: string;
}

export interface IQAgentResponse {
  status: 'success' | 'error' | 'processing';
  timestamp: string;
  summary: ComplianceSummary;
  artifacts: ComplianceArtifacts;
  graph_context: GraphContext;
  evidence: ComplianceEvidence;
  next_actions: NextAction[];
  llm_response: string;
}

export interface IQComplianceQueryResponse {
  success: boolean;
  data: IQAgentResponse;
  message?: string;
  processing_time_ms?: number;
}

// ===== MEMORY MANAGEMENT TYPES =====

export interface IQMemoryStoreRequest {
  memory_type: 'compliance_insight' | 'enforcement_case' | 'regulatory_change' | 'risk_assessment' | 'best_practice';
  content: Record<string, any>;
  importance_score: number; // 0.0 to 1.0
  tags?: string[];
}

export interface IQMemoryStoreResponse {
  success: boolean;
  data: {
    memory_id: string;
    status: 'stored' | 'updated' | 'failed';
  };
  message: string;
}

export interface IQMemoryRetrievalRequest {
  query: string;
  max_memories?: number; // 1-50, default 10
  relevance_threshold?: number; // 0.0-1.0, default 0.5
}

export interface MemoryNode {
  id: string;
  memory_type: string;
  content: Record<string, any>;
  timestamp: string;
  importance_score: number;
  access_count: number;
  tags: string[];
  confidence_score: number;
}

export interface IQMemoryRetrievalResponse {
  success: boolean;
  data: {
    query_id: string;
    retrieved_memories: MemoryNode[];
    relevance_scores: number[];
    total_memories_searched: number;
    retrieval_strategy: string;
    confidence_score: number;
  };
  message: string;
}

// ===== GRAPH INITIALIZATION TYPES =====

export interface IQGraphInitializationRequest {
  clear_existing?: boolean;
  load_sample_data?: boolean;
}

export interface IQGraphInitializationResponse {
  success: boolean;
  data: {
    status: 'completed' | 'failed' | 'in_progress';
    timestamp: string;
    nodes_created: Record<string, number>;
    relationships_created: number;
    message: string;
  };
}

// ===== HEALTH CHECK TYPES =====

export interface IQHealthCheckResponse {
  status: 'healthy' | 'degraded' | 'unhealthy';
  neo4j_connected: boolean;
  llm_service_available: boolean;
  memory_system_operational: boolean;
  graph_nodes_count?: number;
  last_successful_query?: string;
  uptime_seconds?: number;
  components: {
    neo4j: 'healthy' | 'unhealthy';
    openai: 'healthy' | 'unhealthy';
    memory_manager: 'healthy' | 'unhealthy';
    [key: string]: string;
  };
}

// ===== AGENTIC RAG TYPES (from Postman collection) =====

export interface AgenticRAGRequest {
  query: string;
  document_types?: ('regulation' | 'guidance' | 'policy' | 'evidence')[];
  max_results?: number;
  framework?: string;
  section?: string;
  context_window?: number;
}

export interface AgenticRAGResponse {
  query: string;
  results: Array<{
    document_id: string;
    title: string;
    content: string;
    relevance_score: number;
    document_type: string;
    framework?: string;
    section?: string;
    metadata: Record<string, any>;
  }>;
  total_results: number;
  processing_time_ms: number;
}

export interface MultiDocumentRAGRequest {
  queries: string[];
  framework: string;
  combine_results: boolean;
  max_results_per_query?: number;
}

// ===== TRUST GRADIENT TYPES =====

export type TrustLevel = 'helper' | 'advisor' | 'partner';

export interface TrustGradientStatus {
  current_level: TrustLevel;
  progression_score: number; // 0.0 to 1.0
  interactions_count: number;
  success_rate: number;
  user_feedback_score: number;
  estimated_progression_time: string;
  next_level_requirements: string[];
}

// ===== STREAMING TYPES =====

export interface IQStreamingResponse {
  type: 'progress' | 'result' | 'error' | 'complete';
  stage?: 'perceiving' | 'planning' | 'acting' | 'learning' | 'remembering' | 'responding';
  progress?: number; // 0-100
  message?: string;
  data?: Partial<IQAgentResponse>;
  error?: string;
}

// ===== UTILITY TYPES =====

export interface ComplianceFramework {
  id: string;
  name: string;
  description: string;
  jurisdiction: string;
  mandatory_controls: number;
  optional_controls: number;
  complexity_score: number;
}

export interface RiskAssessment {
  id: string;
  framework: string;
  overall_risk: 'low' | 'medium' | 'high' | 'critical';
  risk_factors: Array<{
    category: string;
    score: number;
    description: string;
    mitigation: string;
  }>;
  recommendations: string[];
  next_review_date: string;
}

export interface ComplianceGap {
  gap_id: string;
  regulation: {
    code: string;
    name: string;
    jurisdiction: string;
  };
  requirement: {
    id: string;
    title: string;
    description: string;
    risk_level: 'low' | 'medium' | 'high' | 'critical';
    mandatory: boolean;
  };
  gap_severity_score: number;
  impact_assessment: string;
  recommended_actions: ActionPlan[];
}

// ===== ERROR TYPES =====

export interface IQAgentError {
  error_type: 'network' | 'validation' | 'processing' | 'rate_limit' | 'service_unavailable';
  message: string;
  details?: Record<string, any>;
  retry_after?: number; // seconds
  correlation_id?: string;
}

// ===== EXPORT ALL TYPES =====

export type {
  // Core request/response types
  IQComplianceQueryRequest,
  IQComplianceQueryResponse,
  IQAgentResponse,
  
  // Memory management
  IQMemoryStoreRequest,
  IQMemoryStoreResponse,
  IQMemoryRetrievalRequest,
  IQMemoryRetrievalResponse,
  MemoryNode,
  
  // Graph operations
  IQGraphInitializationRequest,
  IQGraphInitializationResponse,
  
  // Health and status
  IQHealthCheckResponse,
  TrustGradientStatus,
  TrustLevel,
  
  // Streaming and progress
  IQStreamingResponse,
  
  // Domain objects
  ComplianceSummary,
  ComplianceArtifacts,
  GraphContext,
  ComplianceEvidence,
  NextAction,
  ActionPlan,
  ComplianceFramework,
  RiskAssessment,
  ComplianceGap,
  
  // Agentic RAG
  AgenticRAGRequest,
  AgenticRAGResponse,
  MultiDocumentRAGRequest,
  
  // Utilities
  BaseResponse,
  IQAgentError
};